### PR TITLE
Cirrus: Fix running Validate task on branches

### PR DESCRIPTION
**Pull request by cevich**
Similar to v3.0 branch PR https://github.com/containers/podman/pull/9062

Followup to dbb9943

Despite skipping the `Smoke` check, it was observed on a *new* branch,
the `validate` task (specifically `git-validation`) will fail.  This
is because:

* `$CIRRUS_LAST_GREEN_CHANGE` will be empty on a new branch.
* `$CIRRUS_BASE_SHA` is always empty for runs triggered by branch-push
* `$EPOCH_TEST_COMMIT` will be set to `YOU_FOUND_A_BUG`.

Fix this by eliminating the `Smoke` task entirely, simplifying all
the `make validate` operations into the `validate` cirrus task.

Also, eliminate the `$CIRRUS_BUILD_ID` value as it's confusing and not
actually used anywhere.  It was formerly used for building VM images,
but this has moved to another repo entirely.

### CODE REVIEW

**Comment by rhatdan** [comment]
LGTM
@edsantiago PTAL
@containers/podman-maintainers PTAL

File: contrib/cirrus/runner.sh
Comment (edsantiago) [inline contrib/cirrus/runner.sh:38]: Why is this being removed?

File: contrib/cirrus/runner.sh
Comment (edsantiago) [inline contrib/cirrus/runner.sh:28]: `make tests-included` should only be run on PRs. I realize there's a new `skip smoke` in `.cirrus.yml` but it might be helpful to include a loud comment here explaining why this is meaningful only in PRs and why it should be skipped on tags/branches.

File: contrib/cirrus/runner.sh
Comment (edsantiago) [inline contrib/cirrus/runner.sh:56]: Oh, no. This is a disaster waiting to happen. Please leave this as it was. It's perfectly fine to remove the `Quick checks` step above. This should not be committed under any circumstances as it is. (I really hope I don't need to spend time explaining why).

**Review by edsantiago — CHANGES_REQUESTED** [review]
Some requests for explanations, and one absolute showstopper.

File: contrib/cirrus/lib.sh
Comment (cevich) [reply contrib/cirrus/lib.sh:92]: I intended the comment above to cover this: `# Discover the value when running on a branch or in`***`get_ci_vm.sh`***

For `get_ci_vm.sh`, most runtime `$CIRRUS_xyz` variables are not set automatically, they have to be discovered at runtime or through `lib.sh` defaults (like `$CIRRUS_BRANCH=$DEST_BRANCH`).  As a service to users, the example hook script renames "origin" to "upstream" before cloning their fork as "origin".  Outside of that context, "origin" will always be correct given [how the clone is setup](https://github.com/containers/podman/blob/65daf2eff891b34d2c5c28ad52763aa222cb7f83/.cirrus.yml#L122-L135).

TBH, I'm not very happy about running git-commands when loading `lib.sh`, but I can't think of another way to do this.  My goal was to allow running any of the VM-based cirrus-ci tasks using `get_ci_vm.sh` + hook.  Perhaps that goal is not practical.  If you have any suggestions, they would be very welcome.

File: contrib/cirrus/lib.sh
Comment (cevich) [reply contrib/cirrus/lib.sh:93]: Yes [there is a default](https://github.com/containers/podman/blob/65daf2eff891b34d2c5c28ad52763aa222cb7f83/contrib/cirrus/lib.sh#L80).  The `get_ci_vm.sh` takes care of scooping up all the global env. vars. in `.cirrus.yml`, including `$DEST_BRANCH`.

File: contrib/cirrus/lib.sh
Comment (cevich) [reply contrib/cirrus/lib.sh:102]: When running under `get_ci_vm.sh` is the only time `$RANDOM` would be used.

File: contrib/cirrus/runner.sh
Comment (cevich) [reply contrib/cirrus/runner.sh:28]: Do you mean, `make tests-included` should only be run ***in CI*** for PRs?

File: contrib/cirrus/runner.sh
Comment (cevich) [reply contrib/cirrus/runner.sh:56]: Sorry, I don't understand your concern.  My intention is to split the `make validate` targets across two tasks:

1. "Smoke" which doesn't require compiling anything, just quickly checks code for lint and formatting stuffs.
2. "Validate" which confirms the binary compilation and does more through early testing.

Compiling takes over 5-minutes, so the idea is to fail as early as possible for items that don't require it.  I could see combining the "Validate" and "Smoke" tasks into one, but the cost would be reduced parallelism and possibly longer feedback.

File: contrib/cirrus/runner.sh
Comment (cevich) [reply contrib/cirrus/runner.sh:38]: > unintended consequences

I did not read any comment about this.

**Comment by cevich** [comment]
@edsantiago hold off on responding please...I played around with simply eliminating "smoke" and I may be a simpler solution...

**Comment by cevich** [comment]
@edsantiago how about this now, I simplified it quite a bit, expanded comments, and confirmed (manually) that `hack/get_ci_vm.sh` works properly when you `source contrib/cirrus/lib.sh` on the VM.

**Comment by cevich** [comment]
Note to me: if this is acceptable, need to remote "Smoke" from the branch-protection list of required checks.

**Comment by cevich** [comment]
Note: Observed ~7min from code push to when `make validate` started running.  Previously this was <2 minutes (but smoke did not run all the validate checks).

File: .cirrus.yml
Comment (edsantiago) [inline .cirrus.yml:105]: As a courtesy to future maintainers, could you remove `smoke` from line 49?

File: .cirrus.yml
Comment (edsantiago) [inline .cirrus.yml:169]: I realize this is copy-paste, and the error is in the original, but please remove this unnecessary duplicate `cd`

File: contrib/cirrus/runner.sh
Comment (edsantiago) [inline contrib/cirrus/runner.sh:41]: Confirmed that [this step executed correctly](https://cirrus-ci.com/task/6100705229406208?command=main#L0)

**Review by edsantiago — COMMENTED** [review]
This is much better, thank you. I'll take a harder look tomorrow morning, my brain is tired now.

I see CI complaining about "Smoke Test", and assume that's what you mean by your "protection" comment, and that you know how to deal with that.

**Comment by edsantiago** [comment]
New task-map, with `smoke` removed. 
cirrus-map-pr9063

**Review by saschagrunert — APPROVED** [review]
LGTM

File: .cirrus.yml
Comment (cevich) [reply .cirrus.yml:105]: Of course, good idea.

File: .cirrus.yml
Comment (cevich) [reply .cirrus.yml:169]: done

File: contrib/cirrus/runner.sh
Comment (cevich) [reply contrib/cirrus/runner.sh:41]: oh...that reminds me.  We can override any of the `$CIRRUS_xyz` variables under the global `env:` section in `.cirrus.yml`.  That means it's possible to "fake" a branch or tag push in a PR (temporarily).  I'll do that next, so we have a proper test of all code-paths...

**Comment by cevich** [comment]
Oh! Interesting, I'm glad I tested the branch-push code path.  [Somehow the VM doesn't know what 'origin/master' is](https://cirrus-ci.com/task/6482953154854912):

```
$GOSRC/$SCRIPT_BASE/setup_environment.sh
fatal: ambiguous argument 'origin/master': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'

Exit status: 128
```

@edsantiago do you understand what's going on here (when loading `lib.sh`)?  It must have something to with [how we populate the repository when running because of branch-push or tag-push.](https://github.com/containers/podman/blob/62680c4ddc7e9e4c5db932a48f9878c9bea7527f/.cirrus.yml#L157-L169)

**Comment by cevich** [comment]
Ahh...wait a sec...this might simply be due to 'Check Automation' and 'Ext. services' use the default Cirrus-CI clone code, not our custom script...

**Comment by cevich** [comment]
...alright, let's try just moving the custom clone, so all jobs use it.

**Comment by cevich** [comment]
Ref: [Successful Cirrus-CI run w/ simulated branch-push variables](https://cirrus-ci.com/build/5529167754166272)

**Comment by cevich** [comment]
Ready again for a peep (minor changes + rebase).

**Comment by edsantiago** [comment]
@cevich please LMK if/when this is ready for review again. 

If you need to re-push, could you please remove the `smoke` line from https://github.com/containers/podman/blob/efcd48b9668612eb975a299b79e7b6673a5addce/contrib/cirrus/setup_environment.sh#L184

**Comment by edsantiago** [comment]
Ah, looks like I was too late. OK, taking a new look.

**Comment by cevich** [comment]
> please remove the smoke line from

...and there might be more, I honestly didn't search for it very hard (i.e. at all).

File: hack/get_ci_vm.sh
Comment (edsantiago) [inline hack/get_ci_vm.sh:161]: Is this SC2191 necessary after fixing line 164?

File: hack/get_ci_vm.sh
Comment (edsantiago) [inline hack/get_ci_vm.sh:267]: This is beyond the scope of this review, but the above will fail pretty badly if any value in `ENV` has spaces. It's not worth the effort to deal with it properly, but I think a big fat warning (here and/or line 162) would be in order.

File: .cirrus.yml
Comment (edsantiago) [inline .cirrus.yml:197]: This is way above my ability to review. I've spent too long on it already, and I *think* it's okay, but IMO it's a pretty big deal to disable `make validate` on branch actions (whatever those are). I would like double- and triple-checking from @containers/podman-maintainers . 

Team, the question is, can `make validate` safely be skipped in these CI circumstances? As a reminder, that includes: 

 https://github.com/containers/podman/blob/efcd48b9668612eb975a299b79e7b6673a5addce/Makefile#L651

@cevich if you wouldn't mind re-explaining what `branches` means, I think many of our non-CI-wizard members (myself included) would appreciate it.

**Review by edsantiago — APPROVED** [review]
Looks mostly good to me. Three suggestions (two inline, one mentioned earlier re: removing stray `smoke`). One request for deeper review from the team.

File: hack/get_ci_vm.sh
Comment (cevich) [reply hack/get_ci_vm.sh:161]: Yes, this area is more complex than it appears.  Without those disables, we get:

```

In hack/get_ci_vm.sh line 162:
        $(get_env_vars)
        ^-------------^ SC2207: Prefer mapfile or read -a to split command output (or quote to avoid splitting).


In hack/get_ci_vm.sh line 163:
        VM_IMAGE_NAME="$VM_IMAGE_NAME"
        ^------------^ SC2191: The = here is literal. To assign by index, use ( [index]=value ) with no spaces. To keep as literal, quote it.

Did you mean:
        "VM_IMAGE_NAME=""$VM_IMAGE_NAME"


In hack/get_ci_vm.sh line 164:
        UPSTREAM_REMOTE="upstream"
        ^--------------^ SC2191: The = here is literal. To assign by index, use ( [index]=value ) with no spaces. To keep as literal, quote it.

Did you mean:
        "UPSTREAM_REMOTE=""upstream"
```

I'll beef up the comment to explain more fully

File: hack/get_ci_vm.sh
Comment (cevich) [reply hack/get_ci_vm.sh:267]: Yep, and I've experienced it.  I'll take care of this in the beefed up comment on 162, since I need to touch that anyway.

File: .cirrus.yml
Comment (cevich) [reply .cirrus.yml:197]: Correction: This will be skipped on branch-push (i.e. PR merge), creating a new branch,  and on tag-push.

Esp. for tag-push, we want automation to be as simple as possible, for improved automation reliability.   Any non-essential task failures that require code changes to correct...would necessarily need a new tag to be created (potentially repeating most of the release process).  

For a new branch and a branch update...I'm not sure how to distinguish between the two in a 'skip' statement, so they're lumped together.

IMHO, the danger here is caused by nonsequential merging of PRs.  In other words, PRs that pass `validate` individually, but may not if merged in a different order.  However, in this case, I believe we can rely on either the `build` task failing (i.e. in the daily cirrus-cron build) or a future PR (based on the branch) catching the problem - since `validate` would run in that context.

Comment (cevich) [reply .cirrus.yml:197]: (@edsantiago I've just added a comment covering this in `.cirrus.yml`)

File: hack/get_ci_vm.sh
Comment (edsantiago) [reply hack/get_ci_vm.sh:163]: That's weird. Could it be a difference in the version of shellcheck? I ran this on my laptop before making my comment yesterday:
```console
$ contrib/cirrus/shellcheck.sh
WARNING: It does not appear that containers/automation was installed.
         Functionality of most of this library will be negatively impacted
         This contrib/cirrus/lib.sh was loaded by contrib/cirrus/shellcheck.sh
Shellcheck: PASS
$ rpm -q ShellCheck
ShellCheck-0.7.1-3.fc33.x86_64
```
(I wouldn't have made my comment if shellcheck had failed with SC2191) (the above, needless to say, was run on a tree with the `SC2191` removed from `hack/get_ci_vm.sh` but `SC2207` left in place, because that one is really needed).

Anyhow, it's not worth re-pushing for, but it sounds like we have a problem with ShellCheck, like we always do with linters. It might be worth adding a version dump to the failure message in CI some day.

File: .cirrus.yml
Comment (edsantiago) [reply .cirrus.yml:202]: Thanks. I guess it SGTM but I'd still like confirmatory review from @containers/podman-maintainers 

All else LGTM. Thanks.

File: hack/get_ci_vm.sh
Comment (cevich) [reply hack/get_ci_vm.sh:163]: > Could it be a difference in the version of shellcheck?

Oh, that's probably exactly what it is.  In CI we run shellcheck from F33 whereas my laptop is F32.

> but it sounds like we have a problem with ShellCheck, like we always do with linters

The important one is what gets run in CI, but I agree, we gotta take the bad along with the good...at least until the bad exceeds the good :smile:

File: .cirrus.yml
Comment (cevich) [reply .cirrus.yml:202]: nod

**Comment by edsantiago** [comment]
@containers/podman-maintainers could someone PTAL at my question re: the changed `skip`?

I believe this is ready to merge, but I don't fully understand the branch/tag flows and would like a final blessing from someone who does.

**Comment by edsantiago** [comment]
@cevich quick followup, please remember to:

>Note to me: if this is acceptable, need to remove "Smoke" from the branch-protection list of required checks.

**Comment by cevich** [comment]
> need to remove "Smoke" from the branch-protection list

Already done, thanks for the reminder though.

**Pull request merged**
